src/Spice86.Core/Emulator/Devices/ExternalInput/Intel8259Pic.cs

```diff
@@ -0,0 +1,459 @@
+        if (candidate == 8) {
+            return 8;
+        }
+
+        if (IsSpecialMaskModeEnabled || HasHigherPriority(candidate, ActiveIrqLine)) {
+            return candidate;
+        }
+
+        return 8;
+    }
+
+    /// <summary>
+    ///     Updates the lowest-priority slot used to determine subsequent rotation order.
+    /// </summary>
+    /// <param name="value">IRQ index to treat as lowest priority.</param>
+    internal void SetLowestPriorityIrq(byte value) {
+        _lowestPriorityIrq = (byte)(value & 0x07);
+        if (Logger.IsEnabled(LogEventLevel.Debug)) {
+            Logger.Debug("PIC: Lowest priority set to IRQ {Irq}", _lowestPriorityIrq);
+        }
+    }
+
+    /// <summary>
+    ///     Updates the interrupt mask register and re-evaluates IRQs when newly masked bits intersect pending requests.
+    /// </summary>
+    /// <param name="value">Mask value written by the controller.</param>
+    public void SetInterruptMaskRegister(byte value) {
+        byte change = (byte)(InterruptMaskRegister ^ value); // Bits that have changed become 1.
+        InterruptMaskRegister = value;
+        InterruptMaskRegisterInverted = (byte)~value;
+        Logger.Debug("PIC: IMR updated to {Imr:X2} (changed {Changed:X2})", value, change);
+
+        // Test if changed bits are set in irr and are not being served at the moment.
+        // Those bits have an impact on whether the cpu emulation should be paused or not.
+        if ((InterruptRequestRegister & change & InServiceRegisterInverted) != 0) {
+            CheckForIrq();
+        }
+    }
+
+    /// <summary>
+    ///     Re-evaluates pending IRQs after an end-of-interrupt is issued.
+    /// </summary>
+    public void CheckAfterEoi() {
+        // Update the active IRQ as an EOI is likely to change that.
+        UpdateActiveIrqLine();
+        if ((InterruptRequestRegister & InterruptMaskRegisterInverted & InServiceRegisterInverted) != 0) {
+            CheckForIrq();
+        }
+    }
+
+    /// <summary>
+    ///     Recomputes the highest-priority in-service IRQ.
+    /// </summary>
+    private void UpdateActiveIrqLine() {
+        if (InServiceRegister == 0) {
+            ActiveIrqLine = 8;
+            return;
+        }
+
+        for (byte bitIndex = 0, bitMask = 1; bitIndex < 8; bitIndex++, bitMask <<= 1) {
+            if ((InServiceRegister & bitMask) == 0) {
+                continue;
+            }
+
+            ActiveIrqLine = bitIndex;
+            return;
+        }
+    }
+
+    /// <summary>
+    ///     Validates that an IRQ number is within the supported 0-7 range.
+    /// </summary>
+    /// <param name="value">IRQ value to validate.</param>
+    /// <param name="caller">The name of the calling member populated automatically.</param>
+    /// <returns><see langword="true" /> when the value is valid; otherwise <see langword="false" />.</returns>
+    private bool ValidateIrq(byte value, [CallerMemberName] string? caller = null) {
```

Comment:
Are there valid cases where value is >7 and we dont want the emulator to crash because something very wrong happenened?